[Feature] Support for builtin inverted index

srlch opened the pull request:
## What I'm doing:

This PR introduces builtin inverted index including the entire read and write logic. It builtin based on the bitmap index infrastructure.

Fixes #67167

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces a built-in inverted (GIN) index backed by the bitmap index infrastructure and wires it through FE/BE, planner, and scan runtime.
> 
> - Adds builtin inverted index components: `BuiltinInvertedWriter/Reader/Iterator`, `BuiltinPlugin`, and `SimpleAnalyzer`; new `BUILTIN_INVERTED_INDEX` in `segment.proto`; `imp_lib=builtin` support and plugin factory wiring
> - Integrates index into read path: `ColumnReader` loads builtin metadata; `Segment/SegmentIterator` create/use iterators with `IndexReadOptions`; `LakeConnector` propagates `enable_gin_filter` and `prune_column_after_index_filter` and records GIN filter metrics
> - Extends bitmap index APIs for dictionary scans and unions, and optimizes writer (batched adds, ordered flush, configurable dictionary compression)
> - FE changes: enable builtin GIN in shared-data mode (default to builtin), validate `imp_lib` (clucene|builtin), adjust replicated storage behavior with GIN, planner sets new scan-node flags
> - Proto/Thrift and CMake/test updates to support the new index and behaviors
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit e911c3231058da78275ed17f65a083bb6a3d2476. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

wyb commented on be/src/storage/index/inverted/builtin/builtin_inverted_index_iterator.cpp:
```diff
@@ -0,0 +1,208 @@
+
+Status BuiltinInvertedIndexIterator::read_from_inverted_index(const std::string& column_name, const void* query_value,
+                                                              InvertedIndexQueryType query_type,
+                                                              roaring::Roaring* bit_map) {
```
```suggestion
                                                              roaring::Roaring* bitmap) {
```

alvin-celerdata commented:
@cursor review

wyb commented on be/src/storage/index/inverted/builtin/builtin_inverted_index_iterator.cpp:
```diff
@@ -0,0 +1,208 @@
+    return st;
+}
+
+Status BuiltinInvertedIndexIterator::_equal_query(const Slice* search_query, roaring::Roaring* bit_map) {
```
If it’s tokenized at write time, can we still perform an equal query?

wyb commented on be/src/storage/index/inverted/builtin/builtin_inverted_index_iterator.cpp:
```diff
@@ -0,0 +1,208 @@
+    };
+    Slice from_value = Slice(lower.second);
+    size_t search_size = upper.first - lower.first;
+    auto res = _bitmap_itr->seek_dictionary_by_predicate(predicate, from_value, search_size);
```
If it’s tokenized at write time, is the like predicate on dictionary right?

srlch replied:
Yes, it performs like on dictionary

srlch commented on be/src/storage/index/inverted/builtin/builtin_inverted_index_iterator.cpp:
```diff
@@ -0,0 +1,208 @@
+        bool first = true;
+        while (iss >> cur_predicate) {
+            Slice s(cur_predicate);
+            roaring::Roaring roaring;
```
Fixed

srlch commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/IndexAnalyzer.java:
```diff
@@ -172,9 +170,15 @@ public static void checkInvertedIndexValid(Column column, Map<String, String> pr
-            if (!CLUCENE.name().equalsIgnoreCase(impValue)) {
-                throw new SemanticException("Only support clucene implement for now. ");
+            if (!(CLUCENE.name().equalsIgnoreCase(impValue) || BUILTIN.name().equalsIgnoreCase(impValue))) {
+                throw new SemanticException("Only support clucene or builtin implement for now. ");
```
Fixed

srlch committed:
[Feature] Support for builtin inverted index

This PR introduces builtin inverted index including the entire read and write logic.
It builtin based on the bitmap index infrastructure.

Signed-off-by: srlch <linzichao@starrocks.com>
